FaceIterator.h

```diff
@@ -0,0 +1,271 @@
+                        const std::function<void(int,int)>& boundaryFaceHandler = [](int a, int b){},
+                        MPI_Datatype mpit = MPITypeInfer<T>::type()) {
+        iterate(cellHandler, cellHandler, faceHandler, boundaryFaceHandler, mpit);
+    }
+
+    template<typename T, typename S>
+    void iterate(const std::function<T(int,int)>& externalCellHandler,
+                const std::function<S(int,int)>& internalCellHandler,
+                        const std::function<void(int,int,const T&,const S&)>& faceHandler,
+                        const std::function<void(int,int)>& boundaryFaceHandler = [](int a, int b){},
+                        MPI_Datatype mpit = MPITypeInfer<T>::type()) {
+        const auto& realFaceHandler = [&faceHandler, &internalCellHandler](int fid, int cid, const T& tv) {
+            faceHandler(fid, cid, tv, internalCellHandler(fid, cid));
+        };
+        iterate(externalCellHandler, faceHandler, boundaryFaceHandler, mpit);
+    }
+
+    template<typename T>
+    void iterate(const std::function<T(int,int)>& externalCellHandler,
+                        const std::function<void(int,int,const T&)>& faceHandler,
+                        const std::function<void(int,int)>& boundaryFaceHandler = [](int a, int b){},
+                        MPI_Datatype mpit = MPITypeInfer<T>::type()) {
+        iterate_internal(externalCellHandler, faceHandler, boundaryFaceHandler, mpit);
+    }
+
+    const PUML<Topo>& puml() const {
+        return m_puml;
+    }
+
+private:
+    template<typename T>
+    void iterate_internal(const std::function<T(int,int)>& externalCellHandler,
+                        const std::function<void(int,int,const T&)>& faceHandler,
+                        const std::function<void(int,int)>& boundaryFaceHandler,
+                        MPI_Datatype mpit) {
+        int rank = 0, commSize = 1;
+
+        for (int i = 0; i < m_puml.faces().size(); ++i) {
+            const auto& face = m_puml.faces()[i];
+            int lid[2];
+            Upward::cells(m_puml, face, lid);
+            assert(!(lid[0] == -1 && lid[1] != -1));
+
+            if (lid[1] != -1) {
+                const auto gd1 = externalCellHandler(i, lid[1]);
+                faceHandler(i, lid[0], gd1);
+
+                const auto gd0 = externalCellHandler(i, lid[0]);
+                faceHandler(i, lid[1], gd0);
+            }
+            else if (lid[1] == -1 && !face.isShared()) {
+                boundaryFaceHandler(i, lid[0]);
+            }
+        }
+
+#ifdef USE_MPI
+        MPI_Comm_rank(m_puml.comm(), &rank);
+        MPI_Comm_size(m_puml.comm(), &commSize);
+
+        std::vector<T> transferSend(m_transferDisp[commSize]);
+        std::vector<T> transferReceive(m_transferDisp[commSize]);
+        for (int i = 0; i < transferSend.size(); ++i) {
+            transferSend[i] = externalCellHandler(m_transferFace[i], m_transferCell[i]);
+        }
+
+        if (m_sparseComm) {
```

Comment:
I think the reason back then was that I just added this option since PUML wasn't really bound to SeisSol—the reasoning was iirc that if there's some implementation which optimizes an `Alltoall` better than some sends/receives, then this option could be chosen. (or, if we really have a theoretically-badly-distributed mesh where every rank has adjacent elements on every other rank in the comm)

Personally, I'd keep it for now, the one extra line shoudn't hurt—I hope.